### PR TITLE
#45998 Adds new hook interface to support custom hook base classes

### PR DESCRIPTION
**Pull request by josh-t**
This change adds a new, optional `base_class` argument to the `hook.create_hook_instance` method in core. This argument allows the calling code to override the hook base class used when loading the supplied hook paths. Having the ability to set the base class means bundles can define default hook implementations that do not require configuration. 

Additionally, the bulk of the previous logic for `create_hook_instance` has been moved to a new method called `get_hook_class` which simply returns the class without creating an instance. `create_hook_instance` simply calls `get_hook_class` with the required arguments and then returns the instantiated hook. 

The changes to these methods are reflected in the `bundle` convenience methods that wrap them. 

**NOTE**: I have not written tests for these methods yet. If you guys feel comfortable with this approach, I'll crank out the tests.

### CODE REVIEW

File: python/tank/hook.py [hunk 1]
```diff
@@ -631,17 +648,22 @@ def create_hook_instance(hook_paths, parent):
...
-                   app, engine or core object.
+    :param base_class: A python class to use as the base class for the created
+        hook. This will override the default hook base class, ``Hook``.
     :returns: Instance of the hook.
     """
+
     # keep track of the current base class - this is used when loading hooks to dynamically
     # inherit from the correct base.
-    _current_hook_baseclass.value = Hook
+    _current_hook_baseclass.value = base_class or Hook
```
Comment (robblau) [inline python/tank/hook.py]: Worth making sure base_class is a descendent of Hook?

File: python/tank/hook.py [hunk 2]
```diff
@@ -611,14 +611,31 @@ def execute_hook_method(hook_paths, parent, method_name, **kwargs):
     return ret_val
 
 
-def create_hook_instance(hook_paths, parent):
+def create_hook_instance(hook_paths, parent, base_class=None):
```
Comment (robblau) [inline python/tank/hook.py]: Worth carving a path to passing in a base class for execut_hook_method as well?

Comment (josh-t) [reply python/tank/hook.py]: good call. added.

File: python/tank/hook.py (hunk 1 above)
Comment (josh-t) [reply python/tank/hook.py]: done.

File: python/tank/hook.py [hunk 3]
```diff
@@ -586,14 +587,21 @@ def execute_hook_method(hook_paths, parent, method_name, **kwargs):
...
+    while providing a default implementation without the need to configure a
+    base hook.
+
     :param hook_paths: List of full paths to hooks, in inheritance order.
     :param parent: Parent object. This will be accessible inside
                    the hook as self.parent, and is typically an
                    app, engine or core object.
     :param method_name: method to execute. If None, the default method will be executed.
+    :param base_class: A python class to use as the base class for the hook
+        class. This will override the default hook base class, ``Hook``.
```
Comment (manneohrstrom) [inline python/tank/hook.py]: worth pointing out that this needs to derive from `Hook`?

File: python/tank/hook.py [hunk 4]
```diff
@@ -631,17 +656,33 @@ def create_hook_instance(hook_paths, parent):
...
+    :param base_class: A python class to use as the base class for the created
+        hook. This will override the default hook base class, ``Hook``.
     :returns: Instance of the hook.
     """
+
+    if base_class:
+        # ensure the supplied base class is a subclass of Hook
+        if not issubclass(base_class, Hook):
+            raise TankError(
+                "Error retrieving hook class. The supplied base class does not "
```
Comment (manneohrstrom) [inline python/tank/hook.py]: wondering if we could tweak this to be clearer? Could we include the class name of the base class, e.g. "Invalid custom hook base class. The supplied class 'PublishBase' does not inherit from Hook" - why are we printing out the 'hook paths supplied'? it seems unrelated?

File: python/tank/hook.py [hunk 5]
```diff
@@ -631,17 +656,33 @@ def create_hook_instance(hook_paths, parent):
 
         3. /tmp/c.py is finally loaded in, get_hook_baseclass() now returns HookB.
 
-        4. HookC class is instantiated and method method_name is executed.
+        4. HookC class is returned.
```
Comment (manneohrstrom) [inline python/tank/hook.py]: in step 1, can we clarify? 

"get_hook_baseclass() will return Hook at this point" --> "get_hook_baseclass() will return Hook at this point or a custom base class (deriving from hook) if this is passed in via the base_class param"?

File: python/tank/platform/bundle.py [hunk 6]
```diff
@@ -663,11 +663,49 @@ def create_hook_instance(self, hook_expression):
 
         .. note:: For more information about hook syntax, see :class:`~sgtk.Hook`
 
+        An optional `base_class` can be provided to override the default ``Hook``
+        base class. This is useful for bundles that create hook instances at
+        execution time and wish to provide default implementation without the need
+        to configure the base hook.
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: maybe worth pointing our here too that the custom base class needs to derive from hook?

File: python/tank/platform/bundle.py [hunk 7]
```diff
@@ -663,11 +663,49 @@ def create_hook_instance(self, hook_expression):
...
+        resolved_hook_paths = self.__resolve_hook_expression(None, hook_expression)
+        return hook.create_hook_instance(
+            resolved_hook_paths,
+            self,
+            base_class=base_class
+        )
+
+    def get_hook_class(self, hook_expression, base_class=None):
+        """
+        Returns the class for the hook resolved for the given expression.
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: what is the workflow where this is typically needed? An example in the docstring would help greatly i think.

File: python/tank/hook.py (hunk 5 above)
Comment (josh-t) [reply python/tank/hook.py]: done.

File: python/tank/hook.py (hunk 4 above)
Comment (josh-t) [reply python/tank/hook.py]: good call. changed.

File: python/tank/platform/bundle.py (hunk 6 above)
Comment (josh-t) [reply python/tank/platform/bundle.py]: done.

File: python/tank/platform/bundle.py (hunk 7 above)
Comment (josh-t) [reply python/tank/platform/bundle.py]: added.
